### DISCUSSION TITLE
ASDF manager should support "multiple tools on the same line"

### DISCUSSION THREAD

**Discussion by MPV** [renovatebot/renovate#24968] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

Using the ASDF manager, it doesn't (yet) support "multiple tools on the same line", as per:

>**TIP**
>Multiple versions can be set by separating them with a space. For example, to use Python 3.7.2, fallback >to Python 2.7.15 and finally to the system Python, the following line can be added to .tool-versions.
>
>   ```
>   python 3.7.2 2.7.15 system
>   ```
-- https://asdf-vm.com/manage/configuration.html#tool-versions

Whereas the current regex here doesn't seem to implement this...?

https://github.com/renovatebot/renovate/blob/7b025cea48a007faff48cd476a857fe67c4515c9/lib/modules/manager/asdf/extract.ts#L12

For example, I have a `.tool-versions` file with:

```
java semeru-openj9-11.0.19+7_openj9-0.38.0 adoptopenjdk-11.0.19+7 adoptopenjdk-17.0.8+101 temurin-17.0.8+101 adoptopenjdk-8.0.352+8 graalvm-22.0.0.2+java17
```

And this just gives `unsupported-datasource`, as per logs below.

Also, here's a minimal repro issue:
- https://github.com/MPV/renovate-asdf-should-support-one-line-with-multiple-tool-versions/issues/2

### Relevant debug logs

<details><summary>Logs from Renovate job</summary>

```
[...]
          {
            "depName": "java",
            "packageName": "java",
            "skipReason": "unsupported-datasource",
            "updates": []
          },
[...]
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by MPV** [comment]
For example, if we check for newer versions of these tools:
1. the tool versions themselves exist in the index
2. newer versions are available

...as per:
```shell
$ grep java .tool-versions | tr ' ' '\n' | tail -n +2 | xargs -I {} bash -c "asdf list-all java | grep {} -A3"
semeru-openj9-11.0.19+7_openj9-0.38.0
semeru-openj9-11.0.20+8_openj9-0.40.0
semeru-openj9-11.0.20.1+1_openj9-0.40.0
semeru-openj9-16.0.2+7_openj9-0.27.0
adoptopenjdk-11.0.19+7
adoptopenjdk-11.0.20+8
adoptopenjdk-11.0.20+101
adoptopenjdk-12.0.0+33
adoptopenjdk-17.0.8+101
adoptopenjdk-18.0.0+36
adoptopenjdk-18.0.1+10
adoptopenjdk-18.0.2+9
temurin-17.0.8+101
temurin-18.0.0+36
temurin-18.0.1+10
temurin-18.0.2+9
adoptopenjdk-8.0.352+8
adoptopenjdk-8.0.362+9
adoptopenjdk-8.0.372+7
adoptopenjdk-8.0.382+5
graalvm-22.0.0.2+java17
graalvm-22.1.0+java11
graalvm-22.1.0+java17
graalvm-22.2.0+java11
```

**Comment by rarkins** [comment]
Seems an intentional/known limitation: https://github.com/renovatebot/renovate/blob/c8fceec593a184cc7b858bb1186dd52c6a1206ed/lib/modules/manager/asdf/extract.spec.ts#L31-L42

**Reply by MPV** [reply]
@rarkins That makes some sense. Though, do you have any idea why I get `unsupported-datasource` (and not the first version detected, as in the test you linked)?

Thanks for creating the feature request though (as that ultimately would solve the issue, I hope).

**Comment by rarkins** [comment, marked as answer]
Created feature request issue: #24973
